syspaths.py warning message

alan-cugler opened the pull request:
### What does this PR do?
- sets missing variable to `None`
- Created a warning level message:
  - lists missing variables
  - "typo?" prompt
  - shows how & where to explicitly set variables
### What issues does this PR fix or reference?
#52230 
### Previous Behavior
- If variable is missing from _syspaths.py file salt breaks and gives a traceback.
### New Behavior
- Default sets variables not explicitly set in the `_syspaths.py` file
- Prompts user at log warning level about said variables.
### Tests written?
- Was manually tested, no automated tests written.
### Commits signed with GPG?
Yes

alan-cugler committed:
syspaths.py file modified to take missing variables from _syspaths.py and default them to "None"
This will cause variables typoed or missing to be defaulted and allow salt to execute.

alan-cugler committed:
just added a comment for variable name

alan-cugler committed:
Warning log still not working added in elements from observing other salt .py files with log related code.

alan-cugler committed:
Finished flushing out the warning log message.
Added in typo question.
Added in example line to fix the issue.

alan-cugler commented:
I did not add in salt docs reference or URL since said doc doesn't exist yet, and reviewing other log messages showed no such references or URLs set in the messages.

alan-cugler committed:
fixed tab/space mixing and removed trailing white space

alan-cugler committed:
indentation fix

alan-cugler committed:
comment white space pylint test requirement

alan-cugler commented:
I don't get the space after comma error in pylint. later in the same file the lists don't have spaces after commas? pylint test is just bullying me... (T_T)

alan-cugler commented:
@Ch3LL can you explain the error?

Ch3LL commented:
you will want to make sure you point your pylint to the testing pylintrc file. For example I run: `python2 -m pylint --rcfile /home/ch3ll/git/salt/.testing.pylintrc tests/unit/utils/test_stringutils.py`


This should give you similar results as these tests.

But to help point you in the right direction this is how you will need to fix this particular lint error:

```
diff --git a/salt/syspaths.py b/salt/syspaths.py
index 0e61da7639..46c104e9d3 100644
--- a/salt/syspaths.py
+++ b/salt/syspaths.py
@@ -27,7 +27,7 @@ svl = ('ROOT_DIR', 'CONFIG_DIR', 'CACHE_DIR', 'SOCK_DIR',  # salt variable list
     'SRV_ROOT_DIR', 'BASE_FILE_ROOTS_DIR', 'HOME_DIR',
     'BASE_PILLAR_ROOTS_DIR', 'BASE_THORIUM_ROOTS_DIR',
     'BASE_MASTER_ROOTS_DIR', 'LOGS_DIR', 'PIDFILE_DIR',
-    'SPM_PARENT_PATH', 'SPM_FORMULA_PATH','SPM_PILLAR_PATH',
+    'SPM_PARENT_PATH', 'SPM_FORMULA_PATH', 'SPM_PILLAR_PATH',
     'SPM_REACTOR_PATH', 'SHARE_DIR')
 
 try:
```

notice the space after the SPM_FORMULA_PATH comma

alan-cugler committed:
space added to list

alan-cugler commented:
Hey @Ch3LL, I'm not sure why the Centos test failed. This test passed last time, it was the pylint test that had cried wolf before.

alan-cugler committed:
Merge branch 'develop' into syspaths_warning_message

alan-cugler commented:
Cool, not sure why it failed before, but now it passes all the tests!

waynew commented on salt/syspaths.py:
```diff
@@ -21,9 +18,17 @@
-
+typo_warning = False
+log = logging.getLogger(__name__)
+svl = ('ROOT_DIR', 'CONFIG_DIR', 'CACHE_DIR', 'SOCK_DIR',  # salt variable list
```
I think I'd prefer to call this `EXPECTED_VARIABLES`, since it shouldn't be changing. It's also more obvious what `EXPECTED_VARIABLES` means, than coming across `svl` somewhere in your code.

alan-cugler replied:
I can change it to that if you'd like. but I had a comment at the end of the line stating what svl stood for.

waynew replied:
That's how I knew what it meant :)

The reason I prefer having it spelled out is that the other places have the full definition as well. And if you have to put a comment you may as well put it in code.

waynew commented on salt/syspaths.py:
```diff
@@ -32,14 +37,19 @@
+                log.warning('Possible Typo?')
+                log.warning('To dissolve this warning add `[variable] = None` to _syspaths.py')
+            typo_warning = True
+            log.warning('Variable "' + key + '" is missing, value set to None')
```
```suggestion
            log.warning('Variable %r is missing from `_syspaths.py`, value set to `None`', key)
```

waynew commented on salt/syspaths.py:
```diff
@@ -32,14 +37,19 @@
+        else:
+            if not typo_warning:
+                log.warning('Possible Typo?')
+                log.warning('To dissolve this warning add `[variable] = None` to _syspaths.py')
```
Was your thought here that we should only show this typo warning if there's more than one missing value? It seems somewhat odd to me that we'd show it for subsequent errors and not for the first (after all, the first one *could* be a typo).

I think I'd either display it for all or for none - though I'd probably just include it in the other log message.

alan-cugler committed:
1) replace the svl variable,
2) reversed single print for loop mechanic,
3) changed print structure based off of a suggestion by @waynew

Signed-off-by: Alan Cugler <alan.cugler101@gmail.com>

alan-cugler committed:
Merge branch 'syspaths_warning_message' of github.com:terminal-labs/salt into syspaths_warning_message

alan-cugler committed:
Merge branch 'develop' into syspaths_warning_message

alan-cugler commented:
@waynew let me know if anything is still dis-satisfying for the PR. I ended up using `%s` instead of `%r` since the `%r` was printing the unicode with the key variable in the print.

alan-cugler committed:
Merge branch 'develop' into syspaths_warning_message

waynew committed:
Merge branch 'develop' into syspaths_warning_message

waynew commented:
@alan-cugler Looking great! Do you think it would make sense to add a test ensuring that the expected values do get set? I'm not sure if that's a test we *could* automate very well?

alan-cugler commented:
@waynew I am still learning your guy's test suite and how to use the mock library. We can open an issue to add a test in.

alan-cugler committed:
Merge branch 'develop' into syspaths_warning_message

alan-cugler committed:
Merge branch 'develop' into syspaths_warning_message
